File: jdatastudio-api/src/main/java/com/dataserver/api/schema/Field.java
package com.dataserver.api.schema;

import com.dataserver.api.Constants;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Builder;
import lombok.Data;
import lombok.experimental.Tolerate;
import net.sf.cglib.core.CollectionUtils;
import org.mongodb.morphia.annotations.Embedded;

import java.sql.JDBCType;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by gongxinyi on 2018-10-15.
 */
@Data
@Builder
@Embedded
public class Field {
    protected String id;
    protected String name;

    protected String label;
    protected Integer maxLength;
    protected boolean required;
    protected String defaultValue;
    private boolean autoIncremented;
    private boolean partOfPrimaryKey;

    private JDBCType jdbcType;
    private DbColumnType dbColumnType;

    ///////////////////////////jdatastudio 属性/////////////////////////////
    private ComponentType component;
    private boolean showInList;
    private boolean showInShow;
    private boolean showInEdit;
    private boolean showInCreate;
    private boolean showInFilter;
    private boolean alwaysOn;
    private boolean sortable;

    /**
     * 引用字段的属性
     */
    private String reference;
    private String referenceOptionText;
    /**
     * 脱敏类型
     */
    private SensitiveEnum sensitiveType;

    /**
     * 选项值
     */
    private List<ChoiceItem> choices;

    /**
     * 多值过滤
     */
    private boolean multiFilter;

    /**
     * 前端设置下拉选项值
     *
     * @return
     */
    public String getChoicesStr() {
        return choices == null ? "":choices.stream().map(choiceItem ->
                choiceItem.getId() + "|" + choiceItem.getName()
        ).collect(Collectors.joining(Constants.choiceitemtab));
    }

    public void setChoicesStr(String choicesStr){
        choices = new ArrayList<>();
        String[] choicesArray =  choicesStr.split(Constants.choiceitemtab);
        for(String choice:choicesArray){
            String[] choiceItems = choice.split(Constants.choiceitemdelimiter);
            if(choiceItems.length==2){
                choices.add(new ChoiceItem(choiceItems[0],choiceItems[1]));
            }
        }
    }

    /**
     * 是否主字段
     */
    private boolean mainField;

    /**
     * 解决mysql字段名称包含`字符的问题
     *
     * @return
     */
    @JsonProperty("name")
    public String getName() {
        if (name == null) {
            return "";
        } else {
            return name.replace("`", "");
        }
    }

    @Tolerate
    public Field() {
    }
}

File: jdatastudio-admin-app/src/sys/Field.js
import React from "react";
import {
    AutocompleteInput,
    BooleanInput,
    Edit,
    FormDataConsumer,
    FormTab,
    NumberInput,
    ReferenceInput,
    required,
    SelectInput,
    TabbedForm,
    TextInput
} from "react-admin";
import {jdatastudioComponent, sensitiveType} from "../constants";

const redirect = (basePath, id, data) => `/_entity/` + id.substr(0, id.indexOf("_")) + `/fields`;


export const FieldEdit = (props) => (
    <Edit {...props}>
        <TabbedForm redirect={redirect}>
            <FormTab label="字段">
                <TextInput disabled source="name"/>
                <SelectInput source="component" choices={jdatastudioComponent()}/>
                <TextInput source="label" validate={[required()]}/>
                <FormDataConsumer>
                    {({formData, ...rest}) => formData.component === 'Reference' &&
                    <ReferenceInput source="reference" label="引用对象" reference="_entity">
                        <AutocompleteInput optionText="name"/>
                    </ReferenceInput>
                    }
                </FormDataConsumer>
                <FormDataConsumer>
                    {({formData, ...rest}) => formData.component === 'Reference' && formData.reference &&
                    <ReferenceInput source="referenceOptionText" label="显示的字段" reference="_field" filter={{ eid: formData.reference }}>
                        <AutocompleteInput optionValue="name" optionText="label"/>
                    </ReferenceInput>
                    }
                </FormDataConsumer>
                <FormDataConsumer>
                    {({formData, ...rest}) => formData.component === 'Select' &&
                    <TextInput multiline source="choicesStr" label="选项，value|name"/>
                    }
                </FormDataConsumer>
                <BooleanInput source="mainField"/>
                <BooleanInput source="partOfPrimaryKey"/>
                <SelectInput source="sensitiveType" choices={sensitiveType()}/>
            </FormTab>
            <FormTab label="列表">
                <BooleanInput source="showInList"/>
                <BooleanInput source="sortable"/>
                <BooleanInput source="showInShow"/>
            </FormTab>
            <FormTab label="筛选">
                <BooleanInput source="showInFilter"/>
                <BooleanInput source="alwaysOn"/>
                <FormDataConsumer>
                    {({formData, ...rest}) => ['Reference','Select'].includes(formData.component) &&
                    <BooleanInput source="multiFilter" label={"多选过滤"}/>
                    }
                </FormDataConsumer>
            </FormTab>
            <FormTab label="编辑">
                <BooleanInput source="showInEdit"/>
                <BooleanInput source="showInCreate"/>
                <NumberInput source="maxLength"/>
                <BooleanInput source="required"/>
                <TextInput source="defaultValue"/>
                <TextInput disabled source="dbColumnType"/>
            </FormTab>
        </TabbedForm>
    </Edit>

)